sycl/doc/extensions/proposed/sycl_ext_oneapi_raw_kernel.asciidoc

```diff
@@ -0,0 +1,563 @@
+
+If the implementation supports
+link:../experimental/sycl_ext_oneapi_kernel_properties.asciidoc[
+sycl_ext_oneapi_kernel_properties], a raw kernel may be decorated with these
+properties by applying the properties to the function definition as illustrated
+below.
+
+```
+struct my_iota;
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclex::nd_range_kernel<1, my_iota>)
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclex::work_group_size<32>)
+void iota(float start, float *ptr) {
+   // ...
+}
+```
+
+As with standard SYCL kernels, these properties can be queried via
+`kernel::get_info` using either the `info::kernel::attributes` information
+descriptor or the `info::kernel_device_specific` information descriptors.
+
+== Raw kernels that are function templates
+
+A raw kernel may be defined as a function template, however, any instantiation
+that is submitted at runtime must be instantiated at compile time.
+Normally, the application does this with an explicit template instantiation as
+shown below:
+
+```
+template<typename T>
+struct my_iota;
+
+template<typename T>
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclex::range_kernel<1, my_iota<T>>)
+void iota(T start, T *ptr) {
+   // ...
+}
+
+template void iota<int>(int, int *);
+template void iota<float>(float, float *);
+```
+
+As noted earlier, each instantiation must have a unique type-name, as
+illustrated in the example above.
+The application can use this type-name to obtain the `kernel` object for the
+raw kernel:
+
+```
+sycl::kernel myiota_int = exe_bndl.get_kernel<my_iota<int>>();
+sycl::kernel myiota_float = exe_bndl.get_kernel<my_iota<float>>();
+```
+
+
+== Example
+
+The following example demonstrates how to define a raw kernel and then enqueue
+it on a device.
+
+```
+namespace syclex = sycl::ext::oneapi::experimental;
+
+struct my_iota;
+static constexpr size_t NUM = 1024;
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclex::range_kernel<1, my_iota>)
+void iota(float start, float *ptr) {
+  // Get the ID of this kernel iteration.
+  size_t id = syclex::this_kernel::get_id();
+
+  ptr[id] = start + static_cast<float>(id);
+}
+
+void main() {
+  sycl::queue q;
+  sycl::context ctxt = q.get_context();
+
+  // Get a kernel bundle that contains the raw kernel "my_iota".
+  auto exe_bndl =
+    sycl::get_kernel_bundle<my_iota, sycl::bundle_state::executable>(ctxt);
+
+  // Get a kernel object for the "my_iota" function from that bundle.
+  sycl::kernel myiota = exe_bndl.get_kernel<my_iota>();
+
+  float *ptr = sycl::malloc_shared<float>(NUM, q);
+  q.submit([&](sycl::handler &cgh) {
+    // Set the values of the kernel arguments.
+    cgh.set_args(3.14f, ptr);
+
+    cgh.parallel_for({NUM}, myiota);
+  }).wait();
+}
+```
+
+
+== Issues
+
+* The front-end team has expressed concern about implementing this syntax
+  because it requires the front-end to recognize the property names
+  `range_kernel`, `nd_range_kernel`, and `single_task_kernel`.
+  This is necessary because the front-end must know that functions decorated
```

Comment:
With your new proposal would you allow function pointers from a different TU? If not this isn't needed because you know it is device code because the function pointer is passed to get_kernel.

Reply:
I guess we need to support cross TU because online compile is effectively cross TU, correct?

Reply:
Yes, I think we should allow the function pointer to be defined in a separate TU.  Of course, the function still needs to be declared in the TU that has the reference.